# Citation

If you use Bakta in your research, please cite this paper:
> Schwengers O., Jelonek L., Dieckmann M. A., Beyvers S., Blom J., Goesmann A. (2021). Bakta: rapid and standardized annotation of bacterial genomes via alignment-free sequence identification. Microbial Genomics, 7(11). https://doi.org/10.1099/mgen.0.000685

Bakta is *standing on the shoulder of giants* taking advantage of many great software tools and databases. If you find any of these useful for your research, please cite these primary sources, as well.

## Tools

- tRNAscan-SE 2.0 <https://doi.org/10.1093/nar/gkab688>
- Aragorn <https://doi.org/10.1093/nar/gkh152>
- Infernal <https://doi.org/10.1093/bioinformatics/btt509>
- PilerCR <https://doi.org/10.1186/1471-2105-8-18>
- Pyrodigal <https://doi.org/10.21105/joss.04296> Prodigal <https://doi.org/10.1186/1471-2105-11-119>
- Diamond <https://doi.org/10.1038/s41592-021-01101-x>
- BLAST+ <https://doi.org/10.1186/1471-2105-10-421>
- PyHMMER <https://doi.org/10.21105/joss.04296> HMMER <https://doi.org/10.1371/journal.pcbi.1002195>
- AMRFinderPlus <https://doi.org/10.1038/s41598-021-91456-0>
- pyCirclize https://github.com/moshi4/pyCirclize

## Databases

- Rfam: <https://doi.org/10.1002/cpbi.51>
- Mob-suite: <https://doi.org/10.1099/mgen.0.000206>
- DoriC: <https://doi.org/10.1093/nar/gky1014>
- AntiFam: <https://doi.org/10.1093/database/bas003>
- UniProt: <https://doi.org/10.1093/nar/gky1049>
- RefSeq: <https://doi.org/10.1093/nar/gkx1068>
- COG: <https://doi.org/10.1093/bib/bbx117>
- KEGG: <https://doi.org/10.1093/bioinformatics/btz859>
- PHROG: <https://doi.org/10.1093/nargab/lqab067>
- AMRFinder: <https://doi.org/10.1128/AAC.00483-19>
- ISFinder: <https://doi.org/10.1093/nar/gkj014>
- Pfam: <https://doi.org/10.1093/nar/gky995>
- VFDB: <https://doi.org/10.1093/nar/gky1080>